great_expectations/data_context/data_context.py

```diff
@@ -1177,6 +1147,23 @@ def add_datasource(self, name, initialize=True, **kwargs):
             datasource (Datasource)
         """
         logger.debug("Starting BaseDataContext.add_datasource for %s" % name)
+        return self.add_and_initialize_datasource(
+            name=name, initialize=initialize, **kwargs
+        )
+
+    def add_and_initialize_datasource(
+        self, name, initialize=True, **kwargs
```

Comment:
This seems confusing. What would `add_and_initialize_datasource` do with `initialize` set to False?

Reply:
(maybe a clearer question: why are these being split apart?)

Reply:
@jcampbell This is a useful change: they are being split apart in the name of reducing code duplication and increasing code reuse.  The method `add_datasource()` cannot be reused directly -- even if its functionality is wholly reusable, it is decorated with Usage Statistics gathering, which should only run with the `add_datasource()` method verbatim.  However, its core functionality, captured in `add_and_initialize_datasource()` is fully reusable in `get_datasource()`, which does not need to be decorated with Usage Statistics (at least it has not been thus far).  This is also caught by tests; the tests reveal if Usage Statistics are called more times than desired (this is how I found out I could not reuse `add_datasource()` directly in `get_datasource()`.

Reply:
As discussed: split this into three methods:
`add_datasource` -> datasource name and config; adds to project config. Does *not* instantiate, unless flag is set to True (the default)
`get_datasource`: datasource name -> datasource; if not in the cache, then instantiate it
`_instantiate_datasource_from_config`: datasource config -> datasource, also adds to "cached_datasources"